lib/cryptopunk/crypto/tron/checksum_encoding.ex

```diff
@@ -0,0 +1,32 @@
+defmodule Cryptopunk.Crypto.Tron.ChecksumEncoding do
+  @moduledoc false
+
+  @address_length 34
+
+  @spec valid?(String.t()) :: boolean()
+  def valid?(address) do
+    String.starts_with?(address, "T") && String.length(address) == @address_length &&
+      do_check_address_checksum(address)
+  end
+
+  defp sha256(str) do
+    :crypto.hash(:sha256, str)
+  end
+
+  defp do_check_address_checksum(address) do
+    with {:ok, decoded} <-
+           ExBase58.decode(address) do
+      <<address_without_checksum::binary-size(byte_size(decoded) - 4), checksum::binary-size(4)>> =
+        decoded
+
+      double_hash = address_without_checksum |> sha256() |> sha256()
+
+      <<expected_hash::binary-size(4), _::binary>> = double_hash
+
+      expected_hash == checksum
+    else
+      {:error, :invalid_alphabet} -> false
+      {:error, :decode_error} -> false
```

Comment:
what do you think about creating additiona function `validate_address` that will return :ok or {:error, reason}

Reply:
okay 👍🏻

Reply:
@pxp9
[rework checksum encoding to avoid unnecessary with expr and allow imp…](https://github.com/ayrat555/cryptopunk/pull/74/commits/f97524631a62344658b7c2ac80a7ba71d5422423)